cmd/cloud-controller-manager/app/options/options.go

```diff
@@ -18,72 +18,205 @@ package options
 
 import (
 	"fmt"
-	"time"
+	"net"
 
+	"k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
 	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	apiserveroptions "k8s.io/apiserver/pkg/server/options"
 	utilfeature "k8s.io/apiserver/pkg/util/feature"
+	"k8s.io/client-go/kubernetes"
+	clientset "k8s.io/client-go/kubernetes"
+	v1core "k8s.io/client-go/kubernetes/typed/core/v1"
+	restclient "k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/client-go/tools/record"
 	cloudcontrollerconfig "k8s.io/kubernetes/cmd/cloud-controller-manager/app/config"
 	cmoptions "k8s.io/kubernetes/cmd/controller-manager/app/options"
+	"k8s.io/kubernetes/pkg/api/legacyscheme"
+	"k8s.io/kubernetes/pkg/apis/componentconfig"
+	componentconfigv1alpha1 "k8s.io/kubernetes/pkg/apis/componentconfig/v1alpha1"
 	"k8s.io/kubernetes/pkg/master/ports"
-
 	// add the kubernetes feature gates
 	_ "k8s.io/kubernetes/pkg/features"
 
+	"github.com/golang/glog"
 	"github.com/spf13/pflag"
 )
 
 // CloudControllerManagerOptions is the main context object for the controller manager.
 type CloudControllerManagerOptions struct {
-	Generic *cmoptions.GenericControllerManagerOptions
+	CloudProvider     *cmoptions.CloudProviderOptions
+	Debugging         *cmoptions.DebuggingOptions
+	GenericComponent  *cmoptions.GenericComponentConfigOptions
+	KubeCloudShared   *cmoptions.KubeCloudSharedOptions
+	ServiceController *cmoptions.ServiceControllerOptions
+
+	SecureServing *apiserveroptions.SecureServingOptions
+	// TODO: remove insecure serving mode
+	InsecureServing *cmoptions.InsecureServingOptions
+	Authentication  *apiserveroptions.DelegatingAuthenticationOptions
+	Authorization   *apiserveroptions.DelegatingAuthorizationOptions
+
+	Master     string
+	Kubeconfig string
 
 	// NodeStatusUpdateFrequency is the frequency at which the controller updates nodes' status
 	NodeStatusUpdateFrequency metav1.Duration
 }
 
 // NewCloudControllerManagerOptions creates a new ExternalCMServer with a default config.
 func NewCloudControllerManagerOptions() *CloudControllerManagerOptions {
-	componentConfig := cmoptions.NewDefaultControllerManagerComponentConfig(ports.InsecureCloudControllerManagerPort)
+	componentConfig := NewDefaultComponentConfig(ports.InsecureCloudControllerManagerPort)
 
 	s := CloudControllerManagerOptions{
-		// The common/default are kept in 'cmd/kube-controller-manager/app/options/util.go'.
-		// Please make common changes there and put anything cloud specific here.
-		Generic:                   cmoptions.NewGenericControllerManagerOptions(componentConfig),
-		NodeStatusUpdateFrequency: metav1.Duration{Duration: 5 * time.Minute},
+		CloudProvider: &cmoptions.CloudProviderOptions{},
+		Debugging:     &cmoptions.DebuggingOptions{},
+		GenericComponent: &cmoptions.GenericComponentConfigOptions{
+			MinResyncPeriod:         componentConfig.GenericComponent.MinResyncPeriod,
```

Comment:
these assignments are here twice, aren't they? for cloud and kube ctrl mgr. Can we add a constructor?

Reply:
Sure